### ISSUE TITLE
terminal clones with airloop but then cannot be connected to zone

### ISSUE THREAD

**Issue by mleachNREL** [NREL/OpenStudio#4664]
I'm cloning an airloop with a terminal. The terminal and some connections (zone splitter and mixer) come through but it is not connected to any zone. It makes sense that it would not be connected to a zone but the problem is that I can't seem to find a way to connect it to a zone once it has been cloned over. I tried:

1. Using addBranchForZone directly (no error but the terminal didn't get connected to the zone)
2. First using disconnect and then addBranchForZone (same outcome)

To get things to work I had to remove and rebuild the terminal from scratch and then use addBranchForZone 

I've attached the model I get after running clone so you can see what connections come through.

[Terminal_Bug.zip](https://github.com/NREL/OpenStudio/files/9445840/Terminal_Bug.zip)

**Comment by jmarrec** [comment]
I cannot reproduce this. How are you trying to call addBranchForZone on the cloned loop? Are you also trying to supply the cloned ATU? The AirLoopHVACs are somewhat weird if you're used to them...

Basically we have two overloads of AirLoopHVAC::addBranchForZone:

```
    bool addBranchForZone(openstudio::model::ThermalZone& thermalZone)

    bool addBranchForZone(ThermalZone& thermalZone, HVACComponent& airTerminal);
```



The loop also uses what you may call a "connection node". 

(PS: in the future, If you do this with code, please supply some when you open an issue, thanks!)


-----

Here some code that demonstrates that it works fine.

```ruby
include OpenStudio::Model

m = Model.new

a = AirLoopHVAC.new(m)

z = ThermalZone.new(m)
atu = AirTerminalSingleDuctConstantVolumeNoReheat.new(m, m.alwaysOnDiscreteSchedule)

a.addBranchForZone(z, atu)

a.supplyInletNode.setName("#{a.name} Supply Inlet Node")
a.supplyOutletNode.setName("#{a.name} Supply Outlet Node")
if a.mixedAirNode.is_initialized
  a.mixedAirNode.get.setName("#{a.name} Mixed Air Node")
end
a.demandInletNode.setName("#{a.name} Demand Inlet Node")
a.demandOutletNode.setName("#{a.name} Demand Outlet Node")

a.demandMixer.setName("#{a.name} Demand ConnectorMixer")
a.demandSplitter.setName("#{a.name} Demand ConnectorSplitter")

# Rename Zone Air Nodes
m.getThermalZones.each { |z| z.zoneAirNode.setName("#{z.name} Zone Air Node") }

puts "Original Loop Components: #{a.components.map{|c| c.nameString}}"

z2 = ThermalZone.new(m)
a2 = a.clone(m).to_AirLoopHVAC.get()

puts "Clone Loop Components before adding zone: #{a2.components.map{|c| c.nameString}}"
a2.addBranchForZone(z2)
puts "Clone Loop Components after adding zone: #{a2.components.map{|c| c.nameString}}"
puts "z2.equipment:"
puts z2.equipment
```

```ruby
Original Loop Components: ["Air Loop HVAC 1 Supply Inlet Node", "Air Loop HVAC 1 Supply Outlet Node", "Air Loop HVAC 1 Demand Inlet Node", "Air Loop HVAC 1 Demand ConnectorSplitter", "Node 8", "Air Terminal Single Duct Constant Volume No Reheat 1", "Node 5", "Thermal Zone 1", "Node 7", "Air Loop HVAC 1 Demand ConnectorMixer", "Air Loop HVAC 1 Demand Outlet Node"]

Clone Loop Components before adding zone: ["Node 2", "Node 3", "Node 4", "Air Loop HVAC Zone Splitter 1", "Node 10", "Air Terminal Single Duct Constant Volume No Reheat 2", "Node 9", "Air Loop HVAC Zone Mixer 1", "Node 6"]

Clone Loop Components after adding zone: ["Node 2", "Node 3", "Node 4", "Air Loop HVAC Zone Splitter 1", "Node 10", "Air Terminal Single Duct Constant Volume No Reheat 2", "Node 9", "Thermal Zone 2", "Node 11", "Air Loop HVAC Zone Mixer 1", "Node 6"]

z2.equipment:
OS:AirTerminal:SingleDuct:ConstantVolume:NoReheat,
  {4d4c549d-8bd1-4d4a-a25b-9800587065ba}, !- Handle
  Air Terminal Single Duct Constant Volume No Reheat 2, !- Name
  {f99680f9-7263-41d7-b4e0-5c086f88b033}, !- Availability Schedule Name
  {5d50ed5d-3de1-4271-b471-9f107a3b28e6}, !- Air Inlet Node Name
  {73b57f63-3370-4cff-885c-044f5e9fca16}, !- Air Outlet Node Name
  AutoSize;                               !- Maximum Air Flow Rate {m3/s}
```

**Comment by jmarrec** [comment]
The AirLoopHVACs are somewhat weird if you're not used to them...

Basically we have two overloads of AirLoopHVAC::addBranchForZone:

```
    bool addBranchForZone(openstudio::model::ThermalZone& thermalZone)

    bool addBranchForZone(ThermalZone& thermalZone, HVACComponent& airTerminal);
```

First `addBranchForZone(ThermalZone& thermalZone, HVACComponent& airTerminal)` will return false if you pass an airTerminal that is already on an AirLoopHVAC. So if you pass the cloned ATU, which is already on the cloned loop, it will not do anything.

`addBranchForZone(ThermalZone)` will find the last air terminal (if any) and then either:
* If the terminal isn't already connected to a zone (the connection node is right after it), it will connect to it. 
* If the terminal already has a zone, it clones that terminal and attaches to it.

(PS: in the future, **if** you do this with code, please supply some when you open an issue, thanks!)

**Comment by mleachNREL** [comment]
Hi Julien. Sorry for the late response on this. The problem I ran into was that the terminal gets cloned but it is no longer attached to a zone (since that zone is in the model the terminal was cloned from). So the terminal gets cloned over to the new model but it is orphaned because it is no longer connected to its zone and can't be connected to a new one. Ultimately I just had to get the cloned terminal, use relevant info from it to create a new terminal, delete the cloned terminal, and then connect the new terminal to the relevant zone in the model it was cloned to. So I think what I was hoping for would be to clone just the terminal itself and any dependent objects (like a schedule) but none of the broken connections that prevent it from being attached to a zone in the model it was cloned to.
